frontend/src/components/search/Filters/AgencyFilterBody.tsx

```diff
@@ -0,0 +1,122 @@
+import { useSearchParamUpdater } from "src/hooks/useSearchParamUpdater";
+import { QueryContext } from "src/services/search/QueryProvider";
+import {
+  FilterOption,
+  FilterOptionWithChildren,
+} from "src/types/search/searchFilterTypes";
+import { ValidSearchQueryParamData } from "src/types/search/searchQueryTypes";
+import {
+  getAgencyParent,
+  getSiblingOptionValues,
+} from "src/utils/search/searchUtils";
+
+import { useCallback, useContext, useMemo } from "react";
+
+import { AnyOptionCheckbox } from "src/components/search/SearchFilterAccordion/AnyOptionCheckbox";
+import { SearchFilterAccordionProps } from "src/components/search/SearchFilterAccordion/SearchFilterAccordion";
+import SearchFilterCheckbox from "src/components/search/SearchFilterAccordion/SearchFilterCheckbox";
+import SearchFilterSection from "src/components/search/SearchFilterAccordion/SearchFilterSection/SearchFilterSection";
+
+interface AgencyFilterBodyProps extends SearchFilterAccordionProps {
+  referenceOptions?: FilterOption[];
+  topLevelQuery?: Set<string>;
+}
+
+// can rename to NestedCheckboxFilterBody if we can generalize a bit more
+export function AgencyFilterBody({
+  includeAnyOption,
+  title,
+  filterOptions,
+  query,
+  facetCounts,
+  referenceOptions,
+  topLevelQuery,
+}: AgencyFilterBodyProps) {
+  const { queryTerm } = useContext(QueryContext);
+  const { updateQueryParams, setQueryParams } = useSearchParamUpdater();
+
+  const isParentAgencySelected = useCallback(
+    (subAgencyCode: string): boolean => {
+      return topLevelQuery?.has(getAgencyParent(subAgencyCode)) || false;
+    },
+    [topLevelQuery],
+  );
+
+  const toggleOptionChecked = (value: string, isChecked: boolean) => {
+    const newParamValue = new Set(query);
+    isChecked ? newParamValue.add(value) : newParamValue.delete(value);
+    // happy path a normal checkbox experience
+    if (isChecked || !topLevelQuery || !isParentAgencySelected(value)) {
+      return updateQueryParams(newParamValue, "agency", queryTerm);
+    }
+    // handle unchecking a child box when top level parent is selected
+    // need to add children of selected top level, minus the one that has just been removed
+    // since they previously were not present anywhere in the url, but should remain selected
+    const siblingOptions = getSiblingOptionValues(value, filterOptions);
```

Comment:
this is a complex edge case, but without it unchecking a child box when a parent box is selected wouldn't work right